File: buff/classes/354/pub/ch8/BoxConstrained.cc
// g++ -Wall -o BoxConstrained BoxConstrained.cc && ./BoxConstrained

#include <string>
#include <iostream>

using namespace std;

#include "Constraint.h"
#include "Value.h"

template <class Item>
class BoxConstrained {
  Item item;
 public:
  void put(Item item) {
    this->item=item;
  }
  Item get() {
    return item.foo();		// Item is required to provide foo()
  }
};

int main() {
  string s="stuff";
  Value<string> v(s);
  BoxConstrained< Value<string> >* box=
    new BoxConstrained< Value<string> >();
  box->put(v);
  cout << box->get() << endl;
  return 0;
}

File: buff/classes/354/pub/ch8/BoxConstrained.java
// javac BoxConstrained.java && java BoxConstrained

public class BoxConstrained
    <Item extends Constraint<Item>>  {

    private Item item;

    public void put(Item item) {
	this.item=item;
    }

    public Item get() {
	return item.foo();	// Item is required to provide foo()
    }

    public static void main(String[] args) {
	String s="stuff";
	Value<String> v=
	    new Value<String>(s);
	BoxConstrained<Value<String>> box=
	    new BoxConstrained<Value<String>>();
	box.put(v);
	System.out.println(box.get());
    }
	
}
